### DISCUSSION TITLE
HowTo CORS, Trusted Hosts and remote API

### DISCUSSION THREAD

**Discussion by jelovac** [vercel/next.js#33949] [Help]
Hi,

For the past few days I've been trying to figure out how can I send requests to a remote API service on the client side. Whenever I send a request from the client I get the following error in the browser: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSMultipleAllowOriginNotAllowed. 

Since NextJS is running in development mode on a Docker container running NodeJS I am assuming that the headers are set by internal web server which Next.js uses. How can I modify these headers to allow CORS on certain hosts?

I am aware that Next.JS has it's own API which can be used to act as a middle man but I want to avoid this and instead send requests directly to the remote API from the client side.

To get some idea what I am trying to achieve imagine the following. I have a Docker stack which has a container running NextJS in development mode and I have another service which represents the remote API. They are on the same Docker network. And both of them are available publicly using dedicated domains. For the code which is running on NextJS serverside part I am accessing the service using http://internalServiceName and for the code which runs on the NextJS clientside I am accessing the service using the public url of that service. The code which runs server side performs as expected without issues however the client side code runs into CORS issue while accessing the remote service.

Since I do not know how to modify the headers on the web server which is powered when I run "next dev" I tried to solve my issue using proxy with ['async rewrites'](https://nextjs.org/docs/api-reference/next.config.js/rewrites).

```
  async rewrites() {
    return [
      {
        source: '/external-api/:path',
        destination:  'http://myapi.somedomain.com'+ '/:path',
      },
    ]
  },
```

Basically on the code which is executed on the client side I would use /external-api/myroute and I expected this to get rewritten to destination. However when using this method I am getting 500 error from NextJS web server with the following error in the console:  `error - Error: connect ECONNREFUSED 127.0.0.1:80`.

I am currently without ideas, so any advice is welcome.

**Comment by eastwing27** [comment]
I've got the same issue, @jelovac any luck with fixing that?

**Reply by emrecoban** [reply]
As far as I understand, you get CORS policy error. If you fetch data on the server, you don't get any error about CORS. You can use the `'use server'` directive on top of a file, or function, and then call the component, or function in a client component.

**Reply by eastwing27** [reply]
Hi @emrecoban, thanks for the answer!
I get the CORS error when use the local instance of our BE which is external for the Next.js FE app, i.e. we don't use Next.js app server explicitly. The app does work when I use BE staging instance instead. 
Directive `use server` didn't help unfortunately, I've tried to use it within the function that calss axios `get `method
